File: Java/Euler (Java)/src/problems/P001xP050/P004.java
package problems.P001xP050;

import util.Basic;
import base.Problem;

/**
 * P004<br>
 * A palindromic number reads the same both ways. The largest palindrome made<br>
 * from the product of two 2-digit numbers is 9009 = 91 99.<br>
 * Find the largest palindrome made from the product of two 3-digit numbers.<br>
 * Ans: 906609
 */
public class P004 extends Problem
{
   public P004()
   {
      variations = 2;
   }

   @Override
   protected String execute(int variation)
   {
      switch (variation)
      {
         case 0:
            return bruteForce() + "";
         case 1:
         case -1:
            return decreasingSearch() + "";
      }
      return null;
   }

   /**
    * This is the brute force attempt. Just multiply all 3 digit numbers and
    * check if they're palindromic
    * 
    * @return
    */
   private int bruteForce()
   {
      int max = 0;
      for (int n1 = 100; n1 < 1000; n1++)
      {
         for (int n2 = 100; n2 < 1000; n2++)
         {
            int ans = n1 * n2;
            if (Basic.isPalindrome(ans) && ans > max)
               max = ans;
         }
      }
      return max;
   }

   /**
    * Check to see if decreasing numbers are the answer
    * 
    * @return
    */
   private int decreasingSearch()
   {
      int lb = 100 * 100;
      int ub = 999 * 999;

      for (int ii = ub; ii >= lb; ii--)
      {
         // check if it's a palindrome
         if (Basic.isPalindrome(ii))
         {
            // check if it's factorable
            if (isThreesMultiple(ii))
               return ii;
         }
      }
      return 0;
   }

   /**
    * if we find a factor of three digits, the other multiple (because of our
    * bounds) must also be three digits. We start at the square root of the
    * number since it can't be lower and be a multiple.
    * 
    * @param num
    * @return
    */
   private boolean isThreesMultiple(int num)
   {
      int factor = (int) Math.sqrt(num);
      while (factor < 1000)
      {
         if (num / factor <= 100)
            return false; // always less than 3 digits as factor increases
         if (num % factor == 0)
         {
            return true;
         }
         factor++;
      }
      return false;
   }

}

File: OpenCL/Euler (OpenCL)/src/problems/P001xP010/P004.cpp
#include "../../base/problem.h"
#include "../../util/basic.h"

class P004: public Problem
{
	/**
	 * P004<br>
	 * A palindromic number reads the same both ways. The largest palindrome made<br>
	 * from the product of two 2-digit numbers is 9009 = 91 99.<br>
	 * Find the largest palindrome made from the product of two 3-digit numbers.<br>
	 * Ans: 906609
	 */
	public:
		P004()
		{
			problem_number = 4;
			variations = 1;
		}
		~P004() {}

		bool setup(int variation)
		{
			bool retval = false;
			cl = new CLKernel(kernelsize);
			switch (variation)
			{
				case 0:
				case -1:
					retval = brute_setup();
					break;
				default:
					retval = false;
			}
			return retval;
		}

		bool teardown(int variation)
		{
			bool retval = false;
			switch (variation)
			{
				case 0:
				case -1:
					retval = brute_teardown();
					break;
				default:
					retval = false;
			}
			cl->releaseKernel();
			return retval;
		}

		void execute(int variation)
		{
			switch (variation)
			{
				case 0:
				case -1:
					brute();
					break;
				default:
					std::cout << "NOOP!" << std::endl;
					break;
			}
		}

	private:

		/**
		 * Brute force version
		 * Check every single item to see if it's a palindrome!
		 */
		bool brute_setup()
		{
			int initialVal = 0;
			int count = 900*900; // 100->999 * 100->999

			cl->init("src/problems/P001xP010/P004.cl","brute");
         cl->createWriteOnlyMemory(sizeof(int),&answer_mem);

			cl->transferCPUObjectToGPU(sizeof(int),&answer_mem,&initialVal);
			cl->setupOpenCLKernelArgs(1,
					sizeof(cl_mem), &answer_mem);
			cl->waitForQueue(); // sync
			setThreads(count);
			return true;
		}

		bool brute_teardown()
		{
			int ans;
			if (!cl->transferGPUObjectToCPU(sizeof(int), &answer_mem, &ans))
			{
				std::cout << "ERROR:: Could not read answer from GPU!" << std::endl;
				return false;
			}
			answer = to_string(ans);
			clReleaseMemObject(answer_mem);
			return true;
		}

		int brute()
		{
			cl->launchKernel(threadsize);
			cl->waitForQueue();
			return 0;
		}
};
